File: Sources/Sideros/Models/Qwen3Moe.swift
import Foundation
import MLX

public struct Qwen3MoeActivations {
    public let embeddings: MLXArray
    public let blocks: [MLXArray]
    public let norm: MLXArray
    public let logits: MLXArray
}

/// Qwen3's attention — q/k normed per head between the projection and the rotation, no
/// bias — with every MLP replaced by a mixture of 128 experts, 8 per token. The router
/// softmaxes over all of them and renormalizes the eight it keeps; there is no shared
/// expert and no selection bias, so LFM2.5's sigmoid routing does not carry over.
public struct Qwen3Moe: CausalLM {
    private let config: Qwen3MoeConfig
    private let parameters: [String: MLXArray]
    private let quantization: Quantization?
    /// One closure per layer: routing plus the two gathered matmuls.
    let mlps: [(MLXArray) -> MLXArray]
    /// The one-token step, compiled in two chains around the cache: `pre` is
    /// [x, offset] → [queries, keys, values] (norm, qkv, heads, q/k norms, rotation) and
    /// `post` is [attended, x] → [block output] (o_proj, residual, norm, MoE, residual).
    /// The rotary offset rides in as an array — as an op attribute the trace would
    /// freeze it at its first value. Only T=1 compiles: the shape never changes, so it
    /// traces once; the cache stays outside because its shape grows.
    let steps: [(pre: ([MLXArray]) -> [MLXArray], post: ([MLXArray]) -> [MLXArray])]

    public init(
        config: Qwen3MoeConfig, parameters: [String: MLXArray], quantization: Quantization? = nil
    ) {
        self.config = config
        self.quantization = quantization

        var parameters = parameters
        for layer in 0..<config.numHiddenLayers {
            let p = "model.layers.\(layer).self_attn."
            for suffix in ["weight", "scales", "biases"] {
                let parts = ["q_proj.", "k_proj.", "v_proj."].compactMap {
                    parameters[p + $0 + suffix]
                }
                guard parts.count == 3 else { continue }
                // Materialized like gate_up below: `compile` pins captured arrays, and
                // an unevaluated concatenation would re-run inside every traced call.
                let fused = concatenated(parts, axis: 0)
                eval(fused)
                parameters[p + "qkv_proj." + suffix] = fused
            }
        }
        // Gate and up fused row-interleaved ([g₀,u₀,g₁,u₁,…]) so one gather reads both
        // and each (gate, up) pair lands in one simdgroup of the one-token kernel.
        // Row-aligned in all three representations, so the packed weight interleaves
        // exactly like the dense one. Materialized here, originals dropped.
        for layer in 0..<config.numHiddenLayers {
            let p = "model.layers.\(layer).mlp.switch_mlp."
            for suffix in ["weight", "scales", "biases"] {
                let parts = ["gate_proj.", "up_proj."].compactMap { parameters[p + $0 + suffix] }
                guard parts.count == 2 else { continue }
                var shape = parts[0].shape
                shape[1] *= 2
                let fused = stacked(parts, axis: 2).reshaped(shape)
                eval(fused)
                parameters[p + "gate_up_proj." + suffix] = fused
                parameters.removeValue(forKey: p + "gate_proj." + suffix)
                parameters.removeValue(forKey: p + "up_proj." + suffix)
            }
        }
        self.parameters = parameters

        let eps = config.rmsNormEps
        let headDim = config.headDim
        let ropeTheta = config.ropeTheta
        let numHeads = config.numAttentionHeads
        let numKVHeads = config.numKeyValueHeads
        let queryWidth = numHeads * headDim
        let keyValueWidth = numKVHeads * headDim
        let freqs = exp(
            MLXArray(Array(stride(from: Float(0), to: Float(headDim / 2), by: 1)))
                * (-2 / Float(headDim) * log(ropeTheta)))
        let epsArray = MLXArray([eps])
        eval(freqs, epsArray)

        var mlps: [(MLXArray) -> MLXArray] = []
        var steps: [(pre: ([MLXArray]) -> [MLXArray], post: ([MLXArray]) -> [MLXArray])] = []
        for layer in 0..<config.numHiddenLayers {
            let p = "model.layers.\(layer).mlp."
            let router = Linear(p + "gate", parameters: parameters, quantization: quantization)
            let gateUp = Experts(
                p + "switch_mlp.gate_up_proj", parameters: parameters, quantization: quantization)
            let down = Experts(
                p + "switch_mlp.down_proj", parameters: parameters, quantization: quantization)
            let k = config.numExpertsPerTok
            let split = config.numExperts - k
            let hidden = config.hiddenSize
            let inner = config.moeIntermediateSize
            let normTopk = config.normTopkProb

            // The softmax spans all 128 experts, so the eight kept weights depend on the
            // ones dropped — unlike a sigmoid, where selection and weighting are
            // independent. Renormalizing after the cut is what makes the partition
            // equivalent to transformers' sorted topk. `precise` accumulates in float32
            // and rounds back: the normalization runs in the checkpoint's dtype, as in
            // mlx-lm, and float32 here would be a different graph than the reference.
            func route(_ x: MLXArray) -> (chosen: MLXArray, weights: MLXArray) {
                let probs = softmax(router(x), axis: -1, precise: true)
                let chosen = argPartition(probs, kth: split, axis: -1)[0..., 0..., split...]
                var weights = takeAlong(probs, chosen, axis: -1)
                if normTopk {
                    weights = weights / weights.sum(axis: -1, keepDims: true)
                }
                return (chosen, weights)
            }

            func experts(_ tokens: MLXArray, _ indices: MLXArray, sorted: Bool) -> MLXArray {
                let fused = gateUp(tokens, indices: indices, sorted: sorted)
                let pairs = fused.reshaped(Array(fused.shape.dropLast()) + [inner, 2])
                let gated = pairs[.ellipsis, 0]
                let up = pairs[.ellipsis, 1]
                return down(gated * sigmoid(gated) * up, indices: indices, sorted: sorted)
            }

            mlps.append { x in
                let (chosen, weights) = route(x)
                let length = x.dim(-2)
                let routed: MLXArray
                if length * k >= 64 {
                    // Prefill: group the rows by expert so the gather streams each
                    // expert's weight once. A pure reorder — every row is still its own
                    // gemv, so the numbers cannot move.
                    let flat = chosen.reshaped(-1)
                    let order = argSort(flat)
                    let tokens =
                        x.reshaped(length, 1, hidden)[floorDivide(order, MLXArray(Int32(k)))]
                    let out = experts(tokens, flat[order], sorted: true)
                    routed = out[argSort(order)].reshaped(1, length, k, hidden)
                } else {
                    let tokens = expandedDimensions(x, axes: [-2, -3])
                    routed = squeezed(experts(tokens, chosen, sorted: false), axis: -2)
                }
                return (routed * expandedDimensions(weights, axis: -1)).sum(axis: -2)
            }

            let attn = "model.layers.\(layer).self_attn."
            let qkv = Linear(attn + "qkv_proj", parameters: parameters, quantization: quantization)
            let oProj = Linear(attn + "o_proj", parameters: parameters, quantization: quantization)
            let inputNorm = parameters["model.layers.\(layer).input_layernorm.weight"]!
            let postNorm = parameters["model.layers.\(layer).post_attention_layernorm.weight"]!
            let qNorm = parameters[attn + "q_norm.weight"]!
            let kNorm = parameters[attn + "k_norm.weight"]!

            let pre = compile { (inputs: [MLXArray]) -> [MLXArray] in
                let fused = qkv(MLXFast.rmsNorm(inputs[0], weight: inputNorm, eps: eps))
                let (queries, keys) = ropeEpilogue(
                    fused.reshaped(-1), queryHeads: numHeads, kvHeads: numKVHeads,
                    headDim: headDim, qNorm: qNorm, kNorm: kNorm, offset: inputs[1],
                    freqs: freqs, eps: epsArray)
                return [
                    queries.reshaped(1, numHeads, 1, headDim),
                    keys.reshaped(1, numKVHeads, 1, headDim),
                    fused[.ellipsis, (queryWidth + keyValueWidth)...]
                        .reshaped(1, numKVHeads, 1, headDim),
                ]
            }

            let post: ([MLXArray]) -> [MLXArray]
            // `softmaxTopK` renormalizes unconditionally, as both MoE checkpoints here ask
            // it to; a router that keeps the raw softmax weights has to take the op chain.
            if normTopk,
                case .quantized(let guW, let guS, let guB, let guGroup, let guBits) = gateUp,
                case .quantized(let dnW, let dnS, let dnB, let dnGroup, let dnBits) = down,
                moeGemvApplies(
                    hidden: hidden, inner: inner, gateUpGroup: guGroup, downGroup: dnGroup)
            {
                post = compile { (inputs: [MLXArray]) -> [MLXArray] in
                    let projected = oProj(
                        inputs[0].transposed(0, 2, 1, 3).reshaped(1, 1, queryWidth))
                    let (attended, h) = addRmsNorm(
                        inputs[1].reshaped(-1), projected.reshaped(-1), weight: postNorm,
                        eps: epsArray)
                    let (chosen, weights) = softmaxTopK(
                        router(h.reshaped(1, 1, hidden)).reshaped(-1), k: k)
                    let act = moeGateUpAct(
                        h, weight: guW, scales: guS, biases: guB, indices: chosen,
                        groupSize: guGroup, bits: guBits)
                    return [
                        moeDownCombine(
                            act.reshaped(-1), weight: dnW, scales: dnS, biases: dnB,
                            indices: chosen, routing: weights, residual: attended,
                            groupSize: dnGroup, bits: dnBits
                        ).reshaped(1, 1, hidden)
                    ]
                }
            } else {
                post = compile { (inputs: [MLXArray]) -> [MLXArray] in
                    let attended =
                        inputs[1]
                        + oProj(inputs[0].transposed(0, 2, 1, 3).reshaped(1, 1, queryWidth))
                    let h = MLXFast.rmsNorm(attended, weight: postNorm, eps: eps)
                    let (chosen, weights) = route(h)
                    let tokens = expandedDimensions(h, axes: [-2, -3])
                    let routed = squeezed(experts(tokens, chosen, sorted: false), axis: -2)
                    return [
                        attended
                            + (routed * expandedDimensions(weights, axis: -1)).sum(axis: -2)
                    ]
                }
            }
            steps.append((pre: pre, post: post))
        }
        self.mlps = mlps
        self.steps = steps
    }

    /// Total by construction: loadQwen3MoeParameters rejects any checkpoint missing a name.
    private subscript(name: String) -> MLXArray { parameters[name]! }

    private func linear(_ name: String) -> Linear {
        Linear(name, parameters: parameters, quantization: quantization)
    }

    public func embed(_ tokens: MLXArray, offset: Int = 0) -> MLXArray {
        let table = Embedding("model.embed_tokens", parameters: parameters, quantization: quantization)
        return table(tokens)
    }

    public func makeCache() -> [KVCache] { (0..<config.numHiddenLayers).map { _ in KVCache() } }

    public func logits(_ embeddings: MLXArray, cache: [KVCache])
        -> (logits: MLXArray, cache: [KVCache])
    {
        let decoded = decode(embeddings, cache: cache)
        return (decoded.logits, decoded.cache)
    }

    public func callAsFunction(_ ids: [Int32]) -> Qwen3MoeActivations {
        let embeddings = embed(ids)
        let decoded = decode(embeddings, cache: makeCache())
        return Qwen3MoeActivations(
            embeddings: embeddings,
            blocks: decoded.blocks,
            norm: decoded.norm,
            logits: decoded.logits
        )
    }

    private func decode(_ embeddings: MLXArray, cache: [KVCache])
        -> (blocks: [MLXArray], norm: MLXArray, logits: MLXArray, cache: [KVCache])
    {
        var cache = cache
        var x = embeddings
        var blocks: [MLXArray] = []
        // One rotary offset for the whole token: every layer's cache is at the same
        // position, and 48 four-byte uploads per token are 47 too many.
        let offset = x.dim(1) == 1 ? MLXArray([Int32(cache[0].offset)]) : nil
        for layer in 0..<config.numHiddenLayers {
            (x, cache[layer]) = block(x, layer: layer, cache: cache[layer], offset: offset)
            blocks.append(x)
        }

        let normed = rmsNorm(x, prefix: "model.norm")
        let head = config.tieWordEmbeddings ? "model.embed_tokens" : "lm_head"
        return (blocks, normed, linear(head)(normed), cache)
    }

    func block(_ x: MLXArray, layer: Int, cache: KVCache, offset: MLXArray? = nil)
        -> (MLXArray, KVCache)
    {
        if x.dim(1) == 1 {
            let step = steps[layer]
            let projected = step.pre([x, offset ?? MLXArray([Int32(cache.offset)])])
            let (keys, values, grown) = cache.appending(
                keys: projected[1], values: projected[2])
            let attended = MLXFast.scaledDotProductAttention(
                queries: projected[0], keys: keys, values: values,
                scale: 1 / sqrt(Float(config.headDim)), mask: .none)
            return (step.post([attended, x])[0], grown)
        }

        let prefix = "model.layers.\(layer)."
        let (context, grown) = attention(
            rmsNorm(x, prefix: prefix + "input_layernorm"), prefix: prefix, cache: cache)
        let attended = x + context
        return (
            attended
                + mlps[layer](rmsNorm(attended, prefix: prefix + "post_attention_layernorm")),
            grown
        )
    }

    func attention(_ x: MLXArray, prefix: String, cache: KVCache) -> (MLXArray, KVCache) {
        let length = x.dim(1)
        let offset = cache.offset
        let p = prefix + "self_attn."

        let queryWidth = config.numAttentionHeads * config.headDim
        let keyValueWidth = config.numKeyValueHeads * config.headDim
        let fused = linear(p + "qkv_proj")(x)
        let parts = split(fused, indices: [queryWidth, queryWidth + keyValueWidth], axis: -1)

        func heads(_ part: MLXArray, _ count: Int) -> MLXArray {
            part.reshaped(1, length, count, config.headDim).transposed(0, 2, 1, 3)
        }

        let queries = rope(
            rmsNorm(heads(parts[0], config.numAttentionHeads), prefix: p + "q_norm"),
            offset: offset)
        let (keys, values, grown) = cache.appending(
            keys: rope(
                rmsNorm(heads(parts[1], config.numKeyValueHeads), prefix: p + "k_norm"),
                offset: offset),
            values: heads(parts[2], config.numKeyValueHeads))

        let attended = MLXFast.scaledDotProductAttention(
            queries: queries, keys: keys, values: values,
            scale: 1 / sqrt(Float(config.headDim)), mask: length == 1 ? .none : .causal)

        let projected = linear(p + "o_proj")(
            attended.transposed(0, 2, 1, 3).reshaped(1, length, queryWidth))
        return (projected, grown)
    }

    func rmsNorm(_ x: MLXArray, prefix: String) -> MLXArray {
        MLXFast.rmsNorm(x, weight: self[prefix + ".weight"], eps: config.rmsNormEps)
    }

    func rope(_ x: MLXArray, offset: Int = 0) -> MLXArray {
        MLXFast.RoPE(
            x, dimensions: config.headDim, traditional: false, base: config.ropeTheta,
            scale: 1, offset: offset)
    }
}

File: packages/sideros/src/sideros/models/qwen3_moe.py
"""Qwen3-MoE: Qwen3 attention with 128 experts, 8 per token, on every layer.

Property names are the checkpoint's (mlx export layout: experts stacked as
`mlp.switch_mlp.*`). qkv is fused on the output axis and gate‖up row-interleaved
at load, dict-side. The T=1 step routes through the fused kernels when the shapes
tile (`moe_gemv_applies`); prefill gathers sorted by expert — a pure reorder.
"""

import json
import math
from dataclasses import dataclass
from pathlib import Path
from typing import NamedTuple, TypedDict

import mlx.core as mx
import mlx.nn as nn

from sideros.bpe import ByteLevelBPE
from sideros.chat import chat_capabilities
from sideros.checkpoint import (
    checkpoint,
    fuse_qkv,
    interleave_gate_up,
    load_shards,
    prepare_weights,
    stop_tokens,
)
from sideros.core.cache import KVCache
from sideros.core.kernels.add_rms_norm import add_rms_norm, add_rms_norm_applies
from sideros.core.kernels.moe_gemv import moe_down_combine, moe_gate_up_act, moe_gemv_applies
from sideros.core.kernels.moe_route import softmax_topk, softmax_topk_applies
from sideros.core.kernels.rope_epilogue import rope_epilogue, rope_epilogue_applies
from sideros.core.layers import (
    SORTED_GATHER_MIN,
    QuantizedSwitchLinear,
    SwitchLinear,
    sorted_gather,
    split_qkv,
)
from sideros.core.mxcompat import softmax
from sideros.language import LanguageModel, TextLanguageModel
from sideros.model import CompositeModel, ModelInput

# Both default off: measured on the M5 Max (Qwen3-30B-A3B-4bit, 435-token prompt,
# interleaved A/B, median of 5) each kernel LOSES decode here — 150.6 (both on) vs
# 159.9 tok/s (both off), reproduced in a second battery (142.7 vs 152.4). The same
# rope_epilogue wins on the dense 0.6B (310.0 vs 303.3), so the loss is specific to
# this step, not to the kernel. Kept wired for re-measurement.
ROPE_EPILOGUE_KERNEL = False
ADD_RMS_NORM_KERNEL = False


@dataclass(frozen=True)
class Qwen3MoEConfig:
    hidden_size: int
    num_hidden_layers: int
    num_attention_heads: int
    num_key_value_heads: int
    head_dim: int
    vocab_size: int
    rms_norm_eps: float
    rope_theta: float
    tie_word_embeddings: bool
    moe_intermediate_size: int
    num_experts: int
    num_experts_per_tok: int
    norm_topk_prob: bool
    eos_token_id: tuple[int, ...]


class SwitchGLU(nn.Module):
    """Gate and up fused row-interleaved ([g0,u0,g1,u1,…]) at load: one gather reads both."""

    def __init__(self, experts: int, hidden: int, inner: int) -> None:
        super().__init__()
        self.gate_up_proj = SwitchLinear(experts, hidden, 2 * inner)
        self.down_proj = SwitchLinear(experts, inner, hidden)
        self.inner = inner

    def activate(self, fused: mx.array) -> mx.array:
        pairs = fused.reshape(*fused.shape[:-1], self.inner, 2)
        gated = pairs[..., 0]
        return gated * mx.sigmoid(gated) * pairs[..., 1]

    def __call__(self, tokens: mx.array, indices: mx.array, *, sorted_indices: bool) -> mx.array:
        projected = self.gate_up_proj(tokens, indices, sorted_indices=sorted_indices)
        return self.down_proj(self.activate(projected), indices, sorted_indices=sorted_indices)


class Qwen3MoEMLP(nn.Module):
    def __init__(self, config: Qwen3MoEConfig) -> None:
        super().__init__()
        self.gate = nn.Linear(config.hidden_size, config.num_experts, bias=False)
        self.switch_mlp = SwitchGLU(
            config.num_experts, config.hidden_size, config.moe_intermediate_size
        )
        self.hidden = config.hidden_size
        self.k = config.num_experts_per_tok
        self.split = config.num_experts - self.k
        self.norm_topk = config.norm_topk_prob

    def route(self, x: mx.array) -> tuple[mx.array, mx.array]:
        """The softmax spans all experts, so the kept weights depend on the dropped
        ones; renormalizing after the cut matches transformers' sorted topk."""
        probs = softmax(self.gate(x), axis=-1, precise=True)
        chosen = mx.argpartition(probs, kth=self.split, axis=-1)[..., self.split :]
        weights = mx.take_along_axis(probs, chosen, axis=-1)
        if self.norm_topk:
            weights = weights / weights.sum(axis=-1, keepdims=True)
        return chosen, weights

    def __call__(self, x: mx.array) -> mx.array:
        chosen, weights = self.route(x)
        length = x.shape[-2]
        if length * self.k >= SORTED_GATHER_MIN:

            def apply(tokens: mx.array, experts: mx.array) -> mx.array:
                return self.switch_mlp(tokens, experts, sorted_indices=True)

            routed = sorted_gather(x, chosen, k=self.k, hidden=self.hidden, apply=apply)
        else:
            tokens = mx.expand_dims(x, (-2, -3))
            routed = self.switch_mlp(tokens, chosen, sorted_indices=False).squeeze(-2)
        return (routed * mx.expand_dims(weights, -1)).sum(axis=-2)


class Qwen3Attention(nn.Module):
    def __init__(self, config: Qwen3MoEConfig) -> None:
        super().__init__()
        self.heads = config.num_attention_heads
        self.kv_heads = config.num_key_value_heads
        self.head_dim = config.head_dim
        self.rope_theta = config.rope_theta
        self.eps = config.rms_norm_eps
        hidden = config.hidden_size
        queries = self.heads * self.head_dim
        key_values = self.kv_heads * self.head_dim
        self.qkv_proj = nn.Linear(hidden, queries + 2 * key_values, bias=False)
        self.o_proj = nn.Linear(queries, hidden, bias=False)
        self.q_norm = nn.RMSNorm(self.head_dim, eps=config.rms_norm_eps)
        self.k_norm = nn.RMSNorm(self.head_dim, eps=config.rms_norm_eps)

    def _rope(self, x: mx.array, offset: int) -> mx.array:
        return mx.fast.rope(
            x, self.head_dim, traditional=False, base=self.rope_theta, scale=1.0, offset=offset
        )

    def _step_heads(self, x: mx.array, offset: int) -> tuple[mx.array, mx.array, mx.array]:
        """T=1: q/k norm + rotation in one dispatch, v a free slice of the projection."""
        fused = self.qkv_proj(x)
        queries, keys = rope_epilogue(
            fused,
            query_heads=self.heads,
            kv_heads=self.kv_heads,
            head_dim=self.head_dim,
            q_norm=self.q_norm.weight,
            k_norm=self.k_norm.weight,
            offset=offset,
            base=self.rope_theta,
            eps=self.eps,
        )
        values = fused[..., (self.heads + self.kv_heads) * self.head_dim :]
        return (
            queries.reshape(1, self.heads, 1, self.head_dim),
            keys.reshape(1, self.kv_heads, 1, self.head_dim),
            values.reshape(1, self.kv_heads, 1, self.head_dim),
        )

    def __call__(self, x: mx.array, cache: KVCache) -> mx.array:
        length = x.shape[1]
        offset = cache.offset
        query_width = self.heads * self.head_dim
        if ROPE_EPILOGUE_KERNEL and length == 1 and rope_epilogue_applies(self.head_dim):
            queries, rotated, v = self._step_heads(x, offset)
        else:
            q, k, v = split_qkv(
                self.qkv_proj(x),
                heads=self.heads,
                kv_heads=self.kv_heads,
                head_dim=self.head_dim,
            )
            queries = self._rope(self.q_norm(q), offset)
            rotated = self._rope(self.k_norm(k), offset)
        keys, values = cache.update_and_fetch(rotated, v)
        attended = mx.fast.scaled_dot_product_attention(
            queries,
            keys,
            values,
            scale=1 / math.sqrt(self.head_dim),
            mask=None if length == 1 else "causal",
        )
        return self.o_proj(attended.transpose(0, 2, 1, 3).reshape(1, length, query_width))


class Qwen3MoEBlock(nn.Module):
    def __init__(self, config: Qwen3MoEConfig) -> None:
        super().__init__()
        self.self_attn = Qwen3Attention(config)
        self.mlp = Qwen3MoEMLP(config)
        self.input_layernorm = nn.RMSNorm(config.hidden_size, eps=config.rms_norm_eps)
        self.post_attention_layernorm = nn.RMSNorm(config.hidden_size, eps=config.rms_norm_eps)
        self.eps = config.rms_norm_eps
        self.hidden = config.hidden_size
        self.k = config.num_experts_per_tok

    def _fused_step_applies(self) -> bool:
        mlp = self.mlp
        gate_up = mlp.switch_mlp.gate_up_proj
        down = mlp.switch_mlp.down_proj
        return (
            mlp.norm_topk
            and isinstance(gate_up, QuantizedSwitchLinear)
            and isinstance(down, QuantizedSwitchLinear)
            # The gemv kernels read an affine bias per group; MXFP carries none.
            and (gate_up.mode, down.mode) == ("affine", "affine")
            and moe_gemv_applies(
                self.hidden, mlp.switch_mlp.inner, gate_up.group_size, down.group_size
            )
            and softmax_topk_applies(mlp.split + self.k, self.k)
        )

    def _join(self, x: mx.array, cache: KVCache) -> tuple[mx.array, mx.array]:
        """(x + attention, its post-norm). At T=1 one kernel does the pair."""
        if ADD_RMS_NORM_KERNEL and x.shape[1] == 1 and add_rms_norm_applies(self.hidden):
            normed = mx.fast.rms_norm(x, self.input_layernorm.weight, self.eps)
            return add_rms_norm(
                x,
                self.self_attn(normed, cache),
                self.post_attention_layernorm.weight,
                self.eps,
            )
        attended = x + self.self_attn(self.input_layernorm(x), cache)
        return attended, self.post_attention_layernorm(attended)

    def __call__(self, x: mx.array, cache: KVCache) -> mx.array:
        attended, h = self._join(x, cache)
        if x.shape[1] == 1 and self._fused_step_applies():
            gate_up = self.mlp.switch_mlp.gate_up_proj
            down = self.mlp.switch_mlp.down_proj
            assert isinstance(gate_up, QuantizedSwitchLinear)
            assert isinstance(down, QuantizedSwitchLinear)
            assert gate_up.biases is not None and down.biases is not None
            chosen, weights = softmax_topk(self.mlp.gate(h).reshape(-1), self.k)
            act = moe_gate_up_act(
                h.reshape(-1),
                gate_up.weight,
                gate_up.scales,
                gate_up.biases,
                chosen,
                group_size=gate_up.group_size,
                bits=gate_up.bits,
            )
            return moe_down_combine(
                act.reshape(-1),
                down.weight,
                down.scales,
                down.biases,
                chosen,
                weights,
                attended.reshape(-1),
                group_size=down.group_size,
                bits=down.bits,
            ).reshape(1, 1, self.hidden)

        return attended + self.mlp(h)


class Qwen3MoETrunk(nn.Module):
    def __init__(self, config: Qwen3MoEConfig) -> None:
        super().__init__()
        self.embed_tokens = nn.Embedding(config.vocab_size, config.hidden_size)
        self.layers = [Qwen3MoEBlock(config) for _ in range(config.num_hidden_layers)]
        self.norm = nn.RMSNorm(config.hidden_size, eps=config.rms_norm_eps)


class Qwen3MoEActivations(NamedTuple):
    blocks: list[mx.array]
    logits: mx.array


class Qwen3MoE(nn.Module):
    def __init__(self, config: Qwen3MoEConfig) -> None:
        super().__init__()
        self.config = config
        self.model = Qwen3MoETrunk(config)
        if not config.tie_word_embeddings:
            self.lm_head = nn.Linear(config.hidden_size, config.vocab_size, bias=False)

    def make_cache(self) -> list[KVCache]:
        return [KVCache() for _ in self.model.layers]

    def activations(self, ids: mx.array, cache: list[KVCache] | None = None) -> Qwen3MoEActivations:
        cache = cache if cache is not None else self.make_cache()
        x = self.model.embed_tokens(ids)
        blocks: list[mx.array] = []
        for block, layer_cache in zip(self.model.layers, cache, strict=True):
            x = block(x, layer_cache)
            blocks.append(x)
        normed = self.model.norm(x)
        if self.config.tie_word_embeddings:
            logits = self.model.embed_tokens.as_linear(normed)
        else:
            logits = self.lm_head(normed)
        return Qwen3MoEActivations(blocks, logits)

    def __call__(self, ids: mx.array, cache: list[KVCache] | None = None) -> mx.array:
        return self.activations(ids, cache).logits


class _Json(TypedDict):
    model_type: str
    hidden_size: int
    num_hidden_layers: int
    num_attention_heads: int
    num_key_value_heads: int
    head_dim: int
    vocab_size: int
    rms_norm_eps: float
    rope_theta: float
    tie_word_embeddings: bool
    moe_intermediate_size: int
    num_experts: int
    num_experts_per_tok: int
    norm_topk_prob: bool
    eos_token_id: int | list[int]


def _config(path: Path) -> Qwen3MoEConfig:
    raw: _Json = json.loads(path.read_text())
    if raw["model_type"] != "qwen3_moe":
        raise ValueError(f"expected model_type qwen3_moe, got {raw['model_type']!r}")
    eos = raw["eos_token_id"]
    return Qwen3MoEConfig(
        hidden_size=raw["hidden_size"],
        num_hidden_layers=raw["num_hidden_layers"],
        num_attention_heads=raw["num_attention_heads"],
        num_key_value_heads=raw["num_key_value_heads"],
        head_dim=raw["head_dim"],
        vocab_size=raw["vocab_size"],
        rms_norm_eps=raw["rms_norm_eps"],
        rope_theta=raw["rope_theta"],
        tie_word_embeddings=raw["tie_word_embeddings"],
        moe_intermediate_size=raw["moe_intermediate_size"],
        num_experts=raw["num_experts"],
        num_experts_per_tok=raw["num_experts_per_tok"],
        norm_topk_prob=raw["norm_topk_prob"],
        eos_token_id=tuple(eos) if isinstance(eos, list) else (eos,),
    )


def _weights(
    directory: Path, config: Qwen3MoEConfig, dtype: mx.Dtype | None
) -> dict[str, mx.array]:
    layers = config.num_hidden_layers
    return prepare_weights(
        config,
        load_shards(directory),
        [
            lambda weights: fuse_qkv(weights, layers),
            lambda weights: interleave_gate_up(weights, layers),
        ],
        dtype,
    )


def _composite(directory: Path, model: Qwen3MoE) -> LanguageModel[ModelInput]:
    return CompositeModel(
        TextLanguageModel(
            model,
            ByteLevelBPE.from_file(directory / "tokenizer.json"),
            stop=stop_tokens(directory, model.config.eos_token_id),
        ),
        chat_capabilities(directory),
    )


CHECKPOINT = checkpoint((
        "config.json",
        "model*.safetensors",
        "tokenizer.json",
        "tokenizer_config.json",
        "chat_template.jinja",
    ), _config, Qwen3MoE, _weights, _composite)
